### PR TITLE
Generate zsh completions automatically

### PR DESCRIPTION
**Pull request by Rylan12**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [x] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [x] Have you successfully run `brew style` with your changes locally?
- [x] Have you successfully run `brew typecheck` with your changes locally?
- [x] Have you successfully run `brew tests` with your changes locally?
- [x] Have you successfully run `brew man` locally and committed any changes?

-----

Continuation of the work for #10223.

This PR adds the ability for `brew man` to automatically generate zsh completions. Like with the bash completions, this isn't super intelligent at the moment. It doesn't know how to handle conflicting options or anything like that.

One thing that's different for zsh compared to bash completions is that zsh completions allow for descriptions to be used. Descriptions can be used for the brew commands themselves as well as for the options.

To automatically generate the descriptions for options, the `description` parameter defined for the option is used.

To generate descriptions for commands, I chose to use the first sentence of the command description. I skimmed through the generated descriptions to make sure they seemed reasonable but didn't look that closely. The way the first sentence is determined is by splitting the description by periods. There were a few cases where the first period was still within the first sentence of the description (when using `i.e.`/`e.g.` or referring to a file with a dot in the path). I modified those command descriptions to not use periods within the first sentence.

### CODE REVIEW

**Comment by Rylan12** [comment]
One other thing I noticed is that the `completions.rb` file is starting to fill up. At the moment it handles linking/unlinking third-party completions as well as generating completions for two shells (soon to be a third). Each shell seems to need at least two helper methods plus a constant that maps named_arg types to functions.

It may be a good idea to split this up a little bit. One option is to have something like a `CompletionsLinker` and `CompletionsGenerator` module. Additionally or alternatively, the generation for the individual shells could be handled in separate files as well. I know we do something kind of similar with `extend/os` (the difference being that we only want to use one of the os extensions while we'd always want to use all of the shell completions extensions).

Thoughts on the best thing to do here?

File: Library/Homebrew/dev-cmd/sh.rb
Comment (jonchang) [inline Library/Homebrew/dev-cmd/sh.rb:17]:
```suggestion
        Enter an interactive shell for Homebrew's build environment. Use years-battle-hardened
```

I note that most command descriptions start with an imperative verb (display, check, install, remove, etc). We could pattern this after `brew irb` which states:

> Enter the interactive Homebrew Ruby shell.

File: Library/Homebrew/cmd/uses.rb
Comment (EricFromCanada) [inline Library/Homebrew/cmd/uses.rb:24]:
```suggestion
        Show formulae and casks that specify <formula> as a dependency; that is, show dependents
```

**Comment by Rylan12** [comment]
Okay, good news! I have fish completions working locally as well. Those require the changes in this PR but I'm going to hold off on pushing those, for now, to help reduce the size of this (already pretty big) PR.

**Review by zozo2247800 — APPROVED** [review]

**Comment by MikeMcQuaid** [comment]
> Additionally or alternatively, the generation for the individual shells could be handled in separate files as well.

This sounds good to me! Doesn't have to block this PR, though (in fact it may be easier to split up after this is merged). Good thinking about this 🎉

**Review by MikeMcQuaid — APPROVED** [review]

**Comment by Rylan12** [comment]
This will be available to merge later today (once the conflicts are addressed). Should I go ahead and merge or is it better to wait until after https://github.com/Homebrew/brew/pull/10397 is merged per https://github.com/Homebrew/brew/pull/10397#pullrequestreview-575458256. Whichever is merged second will need to be rebased. It doesn't matter to me which one it is.

**Comment by MikeMcQuaid** [comment]
> This will be available to merge later today (once the conflicts are addressed). Should I go ahead and merge or would you rather I wait until after #10397 is merged per [#10397 (review)](https://github.com/Homebrew/brew/pull/10397#pullrequestreview-575458256). Not sure what your estimated timing on that one is. Whichever is merged second will need to be rebased. It doesn't matter to me which one it is.

Let's ship this first. Thanks again @Rylan12, great work!

**Pull request merged**
